src/verifier/step3/Step3Logic.sol

```diff
@@ -0,0 +1,359 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "src/pasta/Pallas.sol";
+import "src/pasta/Vesta.sol";
+import "src/verifier/step2/Step2Data.sol";
+import "src/poseidon/Sponge.sol";
+
+library NIFSPallas {
+    uint256 constant private MOD = Pallas.P_MOD;
+    uint256 constant private NUM_FE_FOR_RO = 24;
+
+    struct NIFS {
+        uint8[32] compressed_comm_T;
+    }
+
+    struct R1CSInstance {
+        Pallas.PallasAffinePoint comm_W;
+        uint256[] X;
+    }
+
+    struct RelaxedR1CSInstance {
+        Pallas.PallasAffinePoint comm_W;
+        Pallas.PallasAffinePoint comm_E;
+        uint256[] X;
+        uint256 u;
+    }
+
+    function verify(
+        NIFS memory nifs,
+        uint256 pp_digest,
+        RelaxedR1CSInstance calldata U1,
+        R1CSInstance calldata U2
+    ) public view returns (RelaxedR1CSInstance memory) {
+        uint256 counter = 0;
+
+        uint256[] memory elementsToHash = new uint256[](NUM_FE_FOR_RO);
+
+        elementsToHash[counter] = pp_digest;
+        counter++;
+
+        //U1.absorb_in_ro
+        // Absorb comm_W
+        elementsToHash[counter] = U1.comm_W.x;
+        counter++;
+        elementsToHash[counter] = U1.comm_W.y;
+        counter++;
+        if (Pallas.isInfinity(U1.comm_W)) {
+            elementsToHash[counter] = 1;
+        }
+        else {
+            elementsToHash[counter] = 0;
+        }
+        counter++;
+
+        // Absorb comm_E
+        elementsToHash[counter] = U1.comm_E.x;
+        counter++;
+        elementsToHash[counter] = U1.comm_E.y;
+        counter++;
+        if (Pallas.isInfinity(U1.comm_E)) {
+            elementsToHash[counter] = 1;
+        }
+        else {
+            elementsToHash[counter] = 0;
+        }
+        counter++;
+
+        // Abosrb u
+        elementsToHash[counter] = U1.u;
+        counter++;
+
+        // Absorb X
+        for (uint256 i = 0; i < U1.X.length; i++) {
+            uint256 entry = U1.X[i];
+            uint256 limb1 = (0x000000000000000000000000000000000000000000000000ffffffffffffffff & entry);
```

Comment:
I think, this limbs extracting would be nice to put into some common utility library, as soon as it is also used in `NIFSVesta`